notebook/notebook/notebook.py

```diff
@@ -131,10 +171,153 @@ def external_url_for(path):
     return url + path
 
 
+def forbidden():
+    return 'Forbidden', 403
+
+
 @app.route('/healthcheck')
 def healthcheck():
     return '', 200
 
+##### New notebook methods meant for consumption from agnostic client#####
+###### Methods that do not call the auth-gateway /verify method must not be directly public #######
+@app.route('/verify/<svc_name>/', methods=['GET'])
+def validate_get(svc_name):
+    access_token = request.args.get('authorization')
+    token = request.args.get('token')
+
+    if not access_token:
+        return forbidden()
+
+    resp = requests.get(f'http://{AUTH_GATEWAY}/verify',
+                        headers={'Authorization': f'Bearer {access_token}'})
+
+    if resp.status_code != 200:
+        return forbidden()
+
+    user_id = resp.headers.get('User')
+
+    if not user_id:
+        return forbidden()
+
+    try:
+        mysql_conn = pymysql.connect(
+        host=IP, port=PORT, user=USER, passwd=PASS, db=DB)
+
+        with mysql_conn.cursor() as cur:
+            cur.execute(f"SELECT svc_name FROM sessions WHERE user_id='{user_id}' AND token='{token}'")
+
+            rows = cur.fetchall()
+
+            if len(rows) != 1 or rows[0][0] != svc_name:
+                return forbidden()
+
+            return '', 200
+    except:
+        log.exception("/validate : GET : failed")
+        return '', 500
+
+
+@app.route('/api', methods=['GET'])
+def new_api_get():
+    user_id = request.headers.get('User')
+
+    if not user_id:
+        return forbidden()
+
+    try:
+        mysql_conn = pymysql.connect(
+        host=IP, port=PORT, user=USER, passwd=PASS, db=DB)
+
+        sql = f"SELECT name, svc_name, token FROM {TABLE} WHERE user_id='{user_id}' AND state<>'{DELETED_STATE}'"
+
+        rows = []
+        with mysql_conn.cursor() as cur:
+            cur.execute(sql)
+            rows = cur.fetchall()
+
+        return jsonify(rows), 200
+    except Exception as e:
+        log.exception(e)
+        return '', 500
+
+
+@app.route('/api/<token>', methods=['DELETE'])
+def delete_notebook(token):
+    user_id = request.headers.get("User")
+
+    # TODO: Is it possible to have a falsy token value and get here?
+    if not user_id or not token:
+        forbidden()
+
+    sqlDel = f"UPDATE {TABLE} SET state='{DELETED_STATE}' WHERE user_id='{user_id}' AND token='{token}'"
+    sqlFetch = f"SELECT svc_name, pod_name FROM {TABLE} WHERE user_id='{user_id}' AND token='{token}'"
+
+    try:
+        mysql_conn = pymysql.connect(
+            host=IP, port=PORT, user=USER, passwd=PASS, db=DB)
+
+        with mysql_conn.cursor() as cur:
+            rows_affected = cur.execute(sqlDel)
+
+            if rows_affected != 1:
+                log.exception(
+                    f'/api/<token> : DELETE : user_id {user_id} has multiple entries for token {token}')
```

Comment:
This is impossible given the primary key. I'm OK with an `assert rows_affected == 1`.

Reply:
True, good catch.

Reply:
A comment on this: The other reason this exists is to do an implicit check on being given the right token for a given user. The token would otherwise need to be explicitly looked up. The error message isn't right.